spec/system/email_change_spec.rb

```diff
@@ -0,0 +1,116 @@
+# frozen_string_literal: true
+
+describe "Changing email", type: :system do
+  fab!(:password) { "mysupersecurepassword" }
+  fab!(:user) { Fabricate(:user, active: true, password: password) }
+  let(:new_email) { "newemail@example.com" }
+
+  before { Jobs.run_immediately! }
+
+  def generate_confirm_link
+    visit "/my/preferences/account"
+
+    email_dropdown = PageObjects::Components::SelectKit.new(".email-dropdown")
+    expect(email_dropdown.visible?).to eq(true)
+    email_dropdown.select_row_by_value("updateEmail")
+
+    find("#change-email").fill_in with: "newemail@example.com"
+
+    find(".save-button button").click
+
+    wait_for(timeout: 5) { ActionMailer::Base.deliveries.count === 1 }
+
+    if user.admin?
+      get_link_from_email(:old)
+    else
+      get_link_from_email(:new)
+    end
+  end
+
+  def get_link_from_email(type)
+    mail = ActionMailer::Base.deliveries.last
+    expect(mail.to).to contain_exactly(type == :new ? new_email : user.email)
+
+    mail.body.to_s[%r{/u/confirm-#{type}-email/\S+}, 0]
+  end
+
+  it "allows regular user to change their email" do
+    sign_in user
+
+    visit generate_confirm_link
+
+    find(".confirm-new-email .btn-primary").click
+
+    expect(page).to have_css(".dialog-body", text: I18n.t("js.user.change_email.confirm_success"))
+    find(".dialog-footer .btn-primary").click
+
+    expect(user.reload.email).to eq(new_email)
+  end
+
+  it "works when user has totp 2fa" do
```

Comment:
It might be nice to add a test case for a security key. We can use `::Selenium::WebDriver::VirtualAuthenticatorOptions.new` in system tests to simulate the credential authentication process.

Reply:
:+1: added